### PR TITLE
"Containerise" the build process

### PR DESCRIPTION
**Pull request by lovell**
Hi John,

This PR adds a Dockerfile to define a container suitable for building the win64 binaries, plus a shell script to make running the build within it as easy as possible.

I had to make a couple of build verification steps optional and add libpng as dependency of cairo and vips to get it all working. I also removed the `transform.patch` file referred to in #1.

Please consider this a starter-for-10 suggestion as I'm happy to make any changes you see fit. I'll need to update the README with details of this alternative approach too if/when you're happy with it.

Cheers,
Lovell

### CODE REVIEW

File: 8.1/jhbuildrc [hunk 1]
```diff
@@ -125,5 +125,5 @@ notrayicon = True
 # whether not to emit notifications through the notification daemon
 nonotify = True 
 
-
-
+# jhbuild uses a CPU-meltingly excessive default parallelism of 2x CPU cores
```
Comment (lovell) [inline 8.1/jhbuildrc]: This comment could be clearer, sorry. "2x CPU cores" means twice the number of CPU cores you have, e.g. a 4x core machine will invoke 8x parallel builds. The use of Docker adds to this CPU hit and my computer (quite rightly) shut itself down when the CPU temperature reached 90C :(

**Comment by jcupitt** [comment]
Oh nice, I've never actually tried docker before, this is great. I've made a docker branch here with your code in:

https://github.com/jcupitt/build-win64/tree/docker

I added some notes to the README and reworded your comment in `jhbuildrc`. 

It'd be good to get some kind of testing in as well. If the container pulled in wine we could test-run `vips.exe`, this would be enough to check that all the necessary DLLs had been packaged, which would be good.

If we put the win32 python into wine, we could potentially run the full vips test suite. Perhaps that's too much. 

I suppose the other wart is the typelib. We'd need to do an additional non-cross-compile build of vips to generate that.

Is there anything else on your TODO for this feature?

**Comment by lovell** [comment]
Thanks John.

Getting 64-bit wine to work with the latest Ubuntu 15.10 is a bit of a headache - see https://bugs.launchpad.net/ubuntu/+source/wine1.6/+bug/1462359

I had to use Ubuntu 15.10 for the base image as its ships with a more recent version of mingw-w64 that uses a version of gcc that contains a bug fix to allow for the compilation of cairo.

**Comment by jcupitt** [comment]
Oh, funny, I didn't have wine problems here, at least  with just running `vips.exe` with no args. I'll investigate.

**Comment by jcupitt** [comment]
I think I might have it (unsure)! I've changed the Dockerfile to be:

```
RUN dpkg --add-architecture i386 && \
  apt-get update && \
  apt-get install -y \
[trim]
          wine \
          wine1.6-amd64
```

So we get wine32 on, and then add 64-bit support. Then running `./vips.exe` directly won't work, since that relies on a kernel thing to register binary interpreters, so we have to change the test to be `wine vips.exe`.

Anyway, fingers crossed, it's running now.

Off-topic, I've added a few more image processing libraries to the "speed and memory use" test. I've seen imlib2 mentioned as a sharp competitor, so you'll be pleased to learn it's a lot slower. There are a couple of other new ones too. ymagine looks like the best vips alternative. 

http://www.vips.ecs.soton.ac.uk/index.php?title=Speed_and_Memory_Use

**Comment by lovell** [comment]
Nice work getting wine to install without dependency errors. I can confirm that the updated `docker` branch works for me.

**Comment by lovell** [comment]
I've merged our docker branches and added an initial Travis CI config to automagically run the build, e.g. https://travis-ci.org/lovell/build-win64/builds/94583555

We could extend this to have Travis post the resultant .zip file somewhere, e.g. an S3 bucket.

**Comment by jcupitt** [comment]
That's a good idea. I don't know anything about S3, but go ahead if you want, it sounds useful.

**Comment by jcupitt** [comment]
I updated the README slightly to put docker builds first and only mention the direct jhbuild route at the end.

I'll try to add typelib generation, this feels like a must-have feature to me, the Python binding won't work without it.

**Comment by lovell** [comment]
I agree with the typelib stuff. I guess it'll need to install the same version of libvips (inside the container definition?) first.

The Github Releases feature might be more suitable than S3 as Travis can push versioned binary artefacts back to Github for all tagged commits - see https://docs.travis-ci.com/user/deployment/releases

As repo owner you'll need to set this up initially - run the following to get started: 

``` sh
gem install travis
travis setup releases
```

**Comment by jcupitt** [comment]
OK, my branch now does a native linux build first to make the typelib.

I messed about with getting win python working under wine, it seems to work, there are notes in the README. Need to find a win64 gi.repository next.

I noticed the .zip has `/lib` and `/lib64`, which probably isn't correct, sigh.

**Comment by jcupitt** [comment]
I've merged your `.travis.yml` and run `travis setup releases`. 

The `/lib` and `/lib64` stuff seems to be OK, as far as I can tell.

Shall I merge back to master? Is there anything you'd like to sort out first?

**Comment by lovell** [comment]
Looks good to me, thanks for all your time tidying up this part-baked idea!

**Comment by jcupitt** [comment]
OK, merged and done, thank you very much for this nice feature @lovell!

**Pull request merged**